// react-docgen-typescript/src/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var nodeUtils_1 = require("./nodeUtils");
var defaultOptions = {
    target: ts.ScriptTarget.Latest,
    module: ts.ModuleKind.CommonJS
};
function getDocumentation(fileName, options) {
    if (options === void 0) { options = defaultOptions; }
    var program = ts.createProgram([fileName], options);
    var checker = program.getTypeChecker();
    var classes = [];
    var interfaces = [];
    var defaultProps = [];
    var sourceFile = program.getSourceFile(fileName);
    var sourceText = sourceFile.text;
    ts.forEachChild(sourceFile, visit);
    function visit(node) {
        if (!isNodeExported(node)) {
            return;
        }
        if (node.kind === ts.SyntaxKind.ClassDeclaration) {
            var classNode = node;
            var symbol = checker.getSymbolAtLocation(classNode.name);
            var typeArguments = nodeUtils_1.navigate(classNode, ts.SyntaxKind.HeritageClause, ts.SyntaxKind.ExpressionWithTypeArguments);
            var list = nodeUtils_1.getFlatChildren(typeArguments)
                .filter(function (i) { return i.kind === ts.SyntaxKind.Identifier; })
                .map(function (i) { return i.text; });
            symbol.exports.defaultProps.declarations.map(function (obj) {
                obj.initializer.properties.map(function (o) {
                    var defaultValue = null;
                    if (o.initializer.text !== undefined) {
                        defaultValue = o.initializer.text.trim();
                    }
                    else if (o.initializer.body !== undefined) {
                        defaultValue = sourceText.substring(o.initializer.body.statements.pos, o.initializer.body.statements.end).trim();
                    }
                    else if (o.initializer !== undefined) {
                        defaultValue = sourceText.substring(o.initializer.pos, o.initializer.end).trim();
                    }
                    if (defaultValue !== null && defaultValue != 'undefined') {
                        defaultProps[o.name.text] = defaultValue;
                    }
                });
            });
            if (list.length > 0) {
                if (list[0] === 'React') {
                    list.shift();
                    list[0] = 'React.' + list[0];
                }
            }
            classes.push({
                name: symbol.name,
                comment: ts.displayPartsToString(symbol.getDocumentationComment()),
                extends: list.length > 0 ? list[0] : null,
                propInterface: list.length > 1 ? list[1] : null,
            });
        }
        if (node.kind === ts.SyntaxKind.InterfaceDeclaration) {
            var interfaceDeclaration = node;
            if (interfaceDeclaration.parent === sourceFile) {
                var symbol = checker.getSymbolAtLocation(interfaceDeclaration.name);
                var type = checker.getTypeAtLocation(interfaceDeclaration.name);
                var members = type.getProperties().map(function (i) {
                    var symbol = checker.getSymbolAtLocation(i.valueDeclaration.name);
                    var prop = i.valueDeclaration;
                    var typeInfo = getType(prop);
                    return {
                        name: i.getName(),
                        text: i.valueDeclaration.getText(),
                        type: typeInfo.type,
                        values: typeInfo.values,
                        isRequired: prop.questionToken === null,
                        comment: ts.displayPartsToString(symbol.getDocumentationComment()).trim(),
                    };
                });
                var interfaceDoc = {
                    name: symbol.getName(),
                    comment: ts.displayPartsToString(symbol.getDocumentationComment()).trim(),
                    members: members,
                };
                interfaces.push(interfaceDoc);
            }
        }
        else if (node.kind === ts.SyntaxKind.ModuleDeclaration) {
            ts.forEachChild(node, visit);
        }
    }
    function isNodeExported(node) {
        return (node.flags & ts.NodeFlags.Export) !== 0 || (node.parent && node.parent.kind === ts.SyntaxKind.SourceFile);
    }
    return {
        classes: classes,
        interfaces: interfaces,
        defaultProps: defaultProps
    };
}
exports.getDocumentation = getDocumentation;
function getType(prop) {
    var unionType = prop.type;
    if (unionType && unionType.types) {
        return {
            type: 'string',
            values: unionType.types.map(function (i) { return i.getText(); }),
        };
    }
    return {
        type: prop.type.getText(),
    };
}
